base/inference.jl

```diff
@@ -1574,13 +1574,21 @@ function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller
         # so need to check whether the code itself is also inferred
         inf = code.inferred
         if !isa(inf, CodeInfo) || (inf::CodeInfo).inferred
-            return code.rettype
+            if code.jlcall_api == 2
+                return abstract_eval_constant(inf)
+            else
+                return code.rettype
+            end
         end
     end
     frame = typeinf_frame(code, true, true, caller)
     frame === nothing && return Any
     frame = frame::InferenceState
-    return widenconst(frame.bestguess)
+    if frame.linfo.jlcall_api == 2
+        return frame.bestguess
```

Comment:
This would be the first convergence edge from post-optimization, so you would possibly need to re-write 'finish'. Although in this case I think you can cheat: since recursion loops won't get jlcall_api 2, you might be able to turn this into a DAG.

The linfo isn't updated unless .inferred and .cached is set in the frame, so it might not be usable.